ruby/ext/google/protobuf_c/storage.c

```diff
@@ -484,9 +565,26 @@ void layout_set(MessageLayout* layout,
                 void* storage,
                 const upb_fielddef* field,
                 VALUE val) {
-  void* memory = ((uint8_t *)storage) +
-      layout->offsets[upb_fielddef_index(field)];
-  if (is_map_field(field)) {
+  void* memory = slot_memory(layout, storage, field);
+  uint32_t* oneof_case = slot_oneof_case(layout, storage, field);
+
+  if (upb_fielddef_containingoneof(field)) {
+    if (val == Qnil) {
+      // Assigning nil to a oneof field clears the oneof completely.
+      *oneof_case = 0;
+      memset(memory, 0, NATIVE_SLOT_MAX_SIZE);
+    } else {
+      // Set the oneof case *first* in case a GC is triggered during
```

Comment:
Hmm, I'm not sure this analysis makes sense to me. You are worried about "val" being collected prematurely? This shouldn't be a risk because it is stored in a variable on the C stack, and the Ruby GC scans the C stack and registers.

On the other hand, I think these risks are real:

1. you set the case indicating that it is a VALUE, but a GC occurs before you actually set the value, causing layout_mark() to follow the previous value (which might be an int64, float, etc) as if it was a VALUE. This is essentially a dereference of a junk value.

2. you set the new case value to a non-VALUE value, causing the previous VALUE stored there to be collected, but then some early exit causes you to bail from actually setting the new value.

So it seems to me that the important invariant is: guarantee that setting of the oneof_case and setting of the value is atomic, with no calls into the Ruby interpreter in between. Calls into the Ruby interpreter are dangerous because they can cause GC and/or exceptions.